### PR TITLE
Place tag detection and replacing "-" with ""

### PR DESCRIPTION
**Pull request by mridulnagpal**
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [*] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [*] code is in uniquely-named feature branch and has no merge conflicts
* [*] PR is descriptively titled
* [ ] PR body includes `fixes #0000`-style reference to original issue #
* [*] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by mridulnagpal** [comment]
@jywarren Please have a look!

**Comment by jywarren** [comment]
Ah, can you link me back to the description? I think we don't want to do this on initial tag creation but only temporarily when attempting geocoding. Let me check the discussion at the planning issue!

File: app/assets/javascripts/tagging.js [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
+  if (tagname.indexOf("place") != -1) {
+    place = tagname.split(":")[1];
+    place.replace("-", " ");
+    alert("This looks like a location");
+  }
 
   var el = $(selector);
 
-  el.find('.tag-input').val(tagname);
+  el.find('.tag-input').val("tagname");
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: I think the quotes are not needed - we want the variable -- but, we could put this in an `else` clause after checking if it's a place. So:

1. if it's a place, start the dialog box and wait until a) we geocode and b) user says OK or Cancel
2. if it's not, just go ahead and submit the tag!

Make sense? 

Good going Mridul, exciting to see this start off!

Comment (mridulnagpal) [reply app/assets/javascripts/tagging.js]: Got it, on it!

**Comment by mridulnagpal** [comment]
@jywarren Made some changes, please have a look

File: Gemfile.lock [hunk 2]
```diff
@@ -63,6 +63,7 @@ GEM
       debug_inspector (>= 0.0.1)
     builder (3.2.3)
     byebug (10.0.2)
+    chronic (0.10.2)
```
Comment (jywarren) [inline Gemfile.lock]: Hi, @mridulnagpal -- are these gems used in your code, or was this added by mistake? I guess we can upgrade the versions without a problem, though!

File: app/assets/javascripts/tagging.js [hunk 3]
```diff
@@ -2,6 +2,14 @@ function addTag(tagname, selector) {
 
   selector = selector || '#tagform';
 
+  if (tagname.indexOf("place") != -1) {
+    place = tagname.split(":")[1];
+    place.replace("-", " ");
+    geocodeStringAndPan(place);
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: do we need to put the prompt code into a callback so it executes /after/ the geocode is complete? Or does this just work?

File: app/assets/javascripts/tagging.js [hunk 4]
```diff
@@ -2,6 +2,14 @@ function addTag(tagname, selector) {
...
+    var choice = prompt("This looks like a location: A) Geocode B) Leave");
+    while(choice == NULL)
+  }
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: I think the `while` may not be needed? the `prompt()` should hang up the process until it's submitted.

File: app/assets/javascripts/tagging.js (hunk 3 above)
Comment (jywarren) [reply app/assets/javascripts/tagging.js]: Yes, perhaps in the `onComplete` parameter!

File: app/controllers/tag_controller.rb [hunk 5]
```diff
@@ -289,8 +289,6 @@ def create
                            uid: current_user.uid,
                            body: " @#{current_user.username} has marked #{tagname.split(':')[1]} as a co-author. ")
 
-        end
```
Comment (jywarren) [inline app/controllers/tag_controller.rb]: Is this a stray deletion? Maybe re-adding the `end` will fix some of the tests?

File: Gemfile.lock (hunk 2 above)
Comment (mridulnagpal) [reply Gemfile.lock]: No, I am not using these gems.

File: app/assets/javascripts/tagging.js (hunk 4 above)
Comment (mridulnagpal) [reply app/assets/javascripts/tagging.js]: Awesome

File: app/controllers/tag_controller.rb (hunk 5 above)
Comment (mridulnagpal) [reply app/controllers/tag_controller.rb]: Sure, on it!!

File: app/assets/javascripts/tagging.js (hunk 3 above)
Comment (mridulnagpal) [reply app/assets/javascripts/tagging.js]: I copied this code from blurred-location library, how do I test all of it?

**Comment by mridulnagpal** [comment]
@jywarren Made the changes, how do we proceed forward?

**Comment by mridulnagpal** [comment]
@jywarren What do we do next, as I don't know how to test it as of yet

**Comment by jywarren** [comment]
Hi, sorry Mridul, I've been traveling and teaching workshops but will take
a look later today! In the meantime can you try asking some other soc folks
for support? Thanks!

**Comment by mridulnagpal** [comment]
@publiclab/soc Any thoughts on how to test this guys?

File: app/assets/javascripts/tagging.js [hunk 6]
```diff
@@ -1,10 +1,19 @@
 function addTag(tagname, selector) {
 
   selector = selector || '#tagform';
+  if (tagname.indexOf("place") != -1) {
+    place = tagname.split(":")[1];
+    place.replace("-", " ");
+
+    confirm("This looks like a location. Is this full description of the location accurate?");
+    if(confirm) {
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: aha - so here, `confirm()` returns  true or false, but doesn't create a `confirm` variable. You can just say:

```js
if (confirm("......")) {
```

Or you could say:

```js
var confirm = confirm("......");
if (confirm) {
```

Make sense?

Comment (mridulnagpal) [reply app/assets/javascripts/tagging.js]: Yes, done!

**Comment by jywarren** [comment]
@mridulnagpal can you take a look at the CodeClimate suggestions too? https://codeclimate.com/github/publiclab/plots2/pull/2734

This did pass though!

File: Gemfile.lock (hunk 2 above)
Comment (jywarren) [reply Gemfile.lock]: Just remember to remove this!

Comment (mridulnagpal) [reply Gemfile.lock]: Done

**Comment by mridulnagpal** [comment]
@jywarren Sure, working on refatoring!!

**Comment by mridulnagpal** [comment]
@publiclab/soc @jywarren Fixed the codeclimate issues, do you know why is this not passing?

**Comment by jywarren** [comment]
Yes, i think it's related to #2593 which we're trying to make progress on!

**Comment by jywarren** [comment]
OK it should work now if you rebase and retrigger it!

**Comment by jywarren** [comment]
Oh oops i also retriggered the build... let's see

File: app/assets/javascripts/tagging.js [hunk 7]
```diff
@@ -1,6 +1,15 @@
 function addTag(tagname, selector) {
 
   selector = selector || '#tagform';
+  if (tagname.indexOf("place") !== -1) {
+    place = tagname.split(":")[1];
+    place.replace("-", " ");
+
+    var confirm = confirm("This looks like a location. Is this full description of the location accurate?");
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: OK, so we have a flow issue here -- think about this, we probably need to do the geocoding /first/ and then present the user with the attempt to determine the full location /after/ it returns, right? 

So if so, we ought to just initiate the geocode automatically. Then, the `confirm()` step ought to go in the onComplete -- that way we can say `Is "____________" an accurate description of the place?`

**Comment by mridulnagpal** [comment]
@jywarren Both checks pass, what next?

**Comment by jywarren** [comment]
Yes, that's just right! Also - if you could please join in community check-ins at #2750 and similar that would be great. It's important other folks hear about your work and they can often offer some help! 👍

**Comment by mridulnagpal** [comment]
@jywarren @publiclab/soc I made some changes, can you guys review? Thanks!!

File: app/assets/javascripts/tagging.js [hunk 8]
```diff
@@ -1,6 +1,18 @@
 function addTag(tagname, selector) {
 
   selector = selector || '#tagform';
+  if (tagname.indexOf("place") !== -1) {
+    place = tagname.split(":")[1];
+    place.replace("-", " ");
+    geo = geocodeStringAndPan(place);
+
+    if (geo.length > 0) {
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: Ok so this block would have to be added in a function as the callback onComplete - or you could move it down inside the onComplete definition on line 102 - make sense?

File: app/assets/javascripts/tagging.js [hunk 9]
```diff
@@ -80,3 +92,19 @@ function initTagForm(deletion_path, selector) {
...
 
 }
+
+function geocodeStringAndPan(string, onComplete) {
+  var url = "https://maps.googleapis.com/maps/api/geocode/json?address=" + string.split(" ").join("+");
+  var Blurred = $.ajax({
+      async: false,
+      url: url
+  });
+  onComplete = onComplete || function onComplete(geometry) {
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: Plus, I believe you'll have to assign the onComplete function to run once this Ajax call is complete.  See jQuery docs to see how to attach this so it's run on the response only. Make sense?

**Comment by mridulnagpal** [comment]
@jywarren Got it, made the changes. Also the onComplete method is called at the end of geocodeStringAndPan, the same way it works in leaflet-blurred-location

File: app/assets/javascripts/tagging.js [hunk 10]
```diff
@@ -80,3 +85,25 @@ function initTagForm(deletion_path, selector) {
...
+    var geo = [lat, lng];
+
+    if (geo.length > 0) {
+      var confirm = confirm("This looks like a location. Is this full description of the location accurate?");
+      if(confirm) {
+      addTag("lat: " + str(geo[0]));
+      addTag("lng: " + str(geo[1]));
+    }    
+  }
+  onComplete(Blurred.responseJSON.results[0].geometry.location);
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: Great but let's attach this to the Ajax call so it runs when they call completes; you'll have to look up how this is done in the jQuery Ajax docs!

File: app/controllers/tag_controller.rb [hunk 11]
```diff
@@ -288,9 +288,8 @@ def create
           node.add_comment(subject: 'co-author',
                            uid: current_user.uid,
                            body: " @#{current_user.username} has marked #{tagname.split(':')[1]} as a co-author. ")
-
         end
-
+
```
Comment (jywarren) [inline app/controllers/tag_controller.rb]: And if possible let's revert this file since there aren't any real changes, no?

Comment (mridulnagpal) [reply app/controllers/tag_controller.rb]: Sure!!

**Comment by mridulnagpal** [comment]
@jywarren Binded the onComplete method with ajax call

File: app/assets/javascripts/tagging.js [hunk 12]
```diff
@@ -80,3 +85,25 @@ function initTagForm(deletion_path, selector) {
   return el;
 
 }
+
+function geocodeStringAndPan(string, onComplete) {
+  var url = "https://maps.googleapis.com/maps/api/geocode/json?address=" + string.split(" ").join("+");
+  var Blurred = $.ajax({
+      async: false,
+      url: url,
+      complete: onComplete(data.responseJSON.results[0].geometry.location),
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: Awesome. Have you confirmed that this works locally? If so, I think this is ready to go!

Comment (mridulnagpal) [reply app/assets/javascripts/tagging.js]: I tried it as a sample in console, but couldn't test the tags as I told you I couldn't figure how to test it. But onComplete runs after the ajax call thats for sure.

**Comment by jywarren** [comment]
What happens when you start plots2 locally and open a page and try to enter
a tag with "place" in it?

Also, we may need to watch that "place" occurs first in the string. I
realized currently isn't it checking for place to occur anywhere, so
"Laplace" would trigger it and possibly error?

**Comment by mridulnagpal** [comment]
image
This is the output, it doesn't do anything. Do you have a clue why?

**Comment by mridulnagpal** [comment]
@jywarren Also corrected the detection of "place" in the tagname

**Comment by jywarren** [comment]
Does this make sense?

**Comment by mridulnagpal** [comment]
@jywarren Got it, I am getting this issue when I try to add 2 tags together.
image

**Comment by jywarren** [comment]
Weird but maybe because in SQLite itbcant respond fast enough for two
entries very fast? Do u have MySQL available?

**Comment by jywarren** [comment]
Try changing your database.yml to use MySQL ( you may need to go through
setup steps again) and see if it works then?

**Comment by mridulnagpal** [comment]
I am stuck on rake db:setup
image

**Comment by mridulnagpal** [comment]
@jywarren I think we have some migration which are not compatible with mysql

**Comment by jywarren** [comment]
Hi, can someone from @publiclab/reviewers help Mridul here? I'm
unfortunately at a meeting this weekend, thanks!

**Comment by jywarren** [comment]
Hi, Mridul - can you check if you've modified your `schema.rb` file at all? Maybe you have to recopy it from `schema.rb.example`?

**Comment by jywarren** [comment]
Retriggered tests -- our upstream fixes made this work now! 

It could be worth pushing to the `unstable` branch to see if it works on mysql. Ask folks in the #2764 check-in for help with this! I think @Gauravano and @icarito wrote up some guidelines in #2750 - give it a try and ask them for help!

File: app/assets/javascripts/tagging.js [hunk 13]
```diff
@@ -80,3 +85,25 @@ function initTagForm(deletion_path, selector) {
...
+  onComplete = onComplete || function onComplete(geometry) {
+    lat = geometry.lat;
+    lng = geometry.lng;
+    
+    var geo = [lat, lng];
+
+    if (geo.length > 0) {
+      var confirm = confirm("This looks like a location. Is this full description of the location accurate?");
+      if(confirm) {
+      addTag("lat: " + str(geo[0]));
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: Ah! you could also do a single `addTag("lat: " + str(geo[0])" + "," + "lng: " + str(geo[1]))` !

Comment (mridulnagpal) [reply app/assets/javascripts/tagging.js]: Awesome!!

**Comment by mridulnagpal** [comment]
@jywarren, I tried testing it and it works, I just need to rewire the form to submit via addTag method

**Comment by mridulnagpal** [comment]
@jywarren @publiclab/soc We are done with the adding of tag, but we need to call addTag() when the form is submitted. Do you know what method is called right now to add Tags to research notes?

File: app/assets/javascripts/tagging.js [hunk 14]
```diff
@@ -80,3 +86,27 @@ function initTagForm(deletion_path, selector) {
...
+        lng = geometry.lng;
+        
+        var geo = [lat, lng];
+
+        if (geo.length > 0) {
+          var r = confirm("This looks like a location. Is this full description of the location accurate?");
+          console.log(geo[0]);
+          console.log(geo[1]);
+          if(r) { 
+            addTag("lat:" + geo[0].toString() + ",lng:" + geo[1].toString()+",place:"+string);
```
Comment (jywarren) [inline app/assets/javascripts/tagging.js]: OK, so I think we can put off refactoring `addTag` for now -- or do it in an independent PR. This should just submit the HTML form. I'm going to merge it!
